samples/qtask_thread.c

```diff
@@ -0,0 +1,344 @@
+        printf("Could not open output file\n");
+        goto end;
+    }
+    //read header, required to resolve the target names to proper ids
+    if (!(in_samhdr = sam_hdr_read(infile))) {
+        printf("Failed to read header from file!\n");
+        goto end;
+    }
+    //start output writer thread for ordered processing
+    twritedata.outfile = outfile2; twritedata.queue = queue2; twritedata.done = &stopcond; twritedata.lock = &stopcondsynch;
+    twritedata.samhdr = in_samhdr;
+    if (pthread_create(&thread, NULL, threadfn_orderedwrite, &twritedata)) {
+        printf("Failed to create writer thread\n");
+        goto end;
+    }
+    //write header
+    if ((sam_hdr_write(outfile1, in_samhdr) == -1) || (sam_hdr_write(outfile2, in_samhdr) == -1)) {
+        printf("Failed to write header\n");
+        goto end;
+    }
+    //check flags and schedule
+    while ((c = sam_read1(infile, in_samhdr, bamdata)) >= 0) {
+        if (bamdata->core.flag & BAM_FREAD1) {
+            //read1, scheduling for unordered output, using Q1; processing occurs in order of thread execution
+            td_unordered *tdata = calloc(1, sizeof(td_unordered));
+            tdata->bamdata = bamdata; tdata->outfile = outfile1; tdata->lock = &filesynch; tdata->samhdr = in_samhdr;
+            if (hts_tpool_dispatch3(pool, queue1, thread_unordered_proc, tdata, (void (*)(void *))&free, (void (*)(void*))&bam_destroy1, 0) == -1) {
+                printf("Failed to schedule unordered processing\n");
+                goto end;
+            }
+        }
+        else if (bamdata->core.flag & BAM_FREAD2) {
+            //read2, scheduling for ordered output, using Q2; proces and result queueing occurs in order of thread execution and result retrieval
+            //showcases the threaded execution and ordered result handling on read2
+            if (hts_tpool_dispatch3(pool, queue2, thread_ordered_proc, bamdata, NULL, (void (*)(void*))&bam_destroy1, 0) == -1) {
+                printf("Failed to schedule ordered processing\n");
+                goto end;
+            }
+        }
+        bamdata = bam_init1();                  //for next read
+    }
+    if (-1 == c) {
+        //EOF read
+        //clear the queues of any tasks; NOTE: will be blocked until queue is cleared
+        if (hts_tpool_process_flush(queue1) == -1 || hts_tpool_process_flush(queue2) == -1) {
+            printf("Failed to flush queues\n");
+            goto end;
+        }
+        ret = EXIT_SUCCESS;
+    }
+    else {
+        printf("Error in reading data\n");
+    }
+
+    //trigger exit for ordered write thread
+    if (thread) {
+        //shutown queues to exit the result wait
+        hts_tpool_process_shutdown(queue1);
+        hts_tpool_process_shutdown(queue2);
+
+        pthread_mutex_lock(twritedata.lock);
+        pthread_cond_signal(twritedata.done);
+        pthread_mutex_unlock(twritedata.lock);
+        pthread_join(thread, NULL);
+        thread = NULL;
```

Comment:
updated to assign integer value.
the join might be required to ensure that the thread completes and exits before the queue destruction in the 'end' section. Another join is there to handle the thread exit in case of errors.

Reply:
`pthread_t` might not be an arithmetic type at all, e.g., it could be a struct; cf https://github.com/samtools/htslib/pull/1153#issuecomment-702937088 et al.

In the far future, [there will be a `PTHREAD_NULL`](https://austingroupbugs.net/view.php?id=599) that you can use in assignments and with `pthread_equal()`, but for now this will need reworking either via a helper variable or by refactoring the cleanup.

Reply:
updated to avoid null/0 assignment/check on pthread_t.